### DISCUSSION TITLE
Coding Style and Naming Convention

### DISCUSSION THREAD

**Discussion by SudoWodo** [SudoWodo/StochOptim#6] [General]
## Reference Style guides
Regarding the coding style is there a specific style guide
that is to be followed for instance :-

1. [Google's R style guide](https://google.github.io/styleguide/Rguide.html)
2. [The tidyverse style guide](https://style.tidyverse.org/index.html) by Hadley Wickham
3.  Somthing else

Google's R style guide was used for Prototype I #5 .

## E-mail 
by John Nash, June 8, 2022

I suggest that the file and function names have same case structure:
e.g., localwrapper or LocalWrapper and use .R rather than .r (the latter
is used by some other structure, I think Eddelbeutel's littler package).

## Discussion - [Remarks on Prototype I #5](https://github.com/SudoWodo/StochOptim/discussions/5) 
by hwborchers, June 8, 2022

**Naming Conventions**

A reasonable convention has been not to start variable and function names with capital letters. Whether to use "localWrapper" or "local_wrapper" or "localwrapper" has always been a question of taste (I prefer the first, but that's only me). Please use the same variation throughout the whole project (and package).

**Comment by hwborchers** [comment]
Well, I'm not sure what the other mentors think, but for me the "Naming conventions" rule of Google's R Style Guide is absolutely unacceptable and is not followed in the R community. (The other rules therein are okay.) The Tidyverse style guide has many reasonable rules and is likely very popular in the Tidyverse (sub)community. Most of these rules are quite normal and are accepted in most programming language communities. I would not ask you to follow the rules in any case, but not using capital lettters for function names seems essential (especially in object-oriented languages).

**Comment by SudoWodo** [comment]
##  Function Names

For function names I have decided to go with `camelCase` (myFunction) eg:-
globalWrapper
localWrapper

##  Class Names
For class name I have decided to stick with `snake_case` this helps me to visually distinguish between function files and class files.
It follows the following format :-

`optimizer_wrapper`

Example :-
DEoptim_wrapper
pso_wrapper
GenSA_wrapper

**Reply by hwborchers** [reply]
That's correct; class names and (sometimes) object names starting with capital letters is done so in most object-oriented systems.

**Reply by SudoWodo** [reply]
sir can you provide example?

**Comment by SudoWodo** [comment]
I feel that the naming scheme i came up with is not convenient and it is not doing what its supposed to do i.e. make names easier to understand for eg :-

1. I found in #10 that a package can have multiple optimizer and their local wrappers will have very complicated names with current naming scheme. Hence `localWrapperadagiosimpleDE` is much more difficult to understand than say `local_wrapper_adagio_simpleDE`.

2.  Some package name start with upper case letter like **DEoptim**  while other start with lower case letter for instance **pso** so with current naming scheme their local wrapper names are `localWrapperDEoptim` and `localWrapperpso` which mixes uppercase and lowercase letters. I think it will be better if snake case is used here as well, so names will become consistent like :- `local_wrapper_DEoptim` and `local_wrapper_pso`

Writing names this way the distinction between the R6Class names and function files names will be lost. But at this moment it seems splitting names with underscore will prove to be more effective overall. Did i miss any scenario which will cause naming blunders in the future? Please let me know.
